### ISSUE TITLE
Cluster samples using PAM and HCA

### ISSUE THREAD

**Issue by lukenoaa** [aomlomics/seusmbon#18]
# Clustering Overview

## Sample Clustering for Ecological Analysis

Assuming you have your **ASV data** and want to cluster the samples based on the ASV abundances, here are the recommended approaches.

### Choose a Dissimilarity Measure

You must first convert your ASV abundance table into a **dissimilarity matrix** (Sample $\times$ Sample).

- **Bray-Curtis Dissimilarity 🏅:** The standard and highly recommended metric for ecological community data. It is robust, ignores double-zeros, and is suitable for proportional abundance data.
- **Aitchison Distance:** Used specifically when data has been **CLR-transformed** (Centered Log-Ratio) to account for its compositional nature.

### Select a Clustering Algorithm

| **Method**                            | **Description**                                              | **Rationale for Metabarcoding**                              |
| ------------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
| **PAM (Partitioning Around Medoids)** | Partitions samples into $k$ groups defined by an actual sample (the **medoid**). | **Preferred partitioning method** as it is more **robust to outliers** and noise than k-means, and works directly with non-Euclidean metrics like Bray-Curtis. |
| **Hierarchical Clustering (HCA)**     | Creates a **dendrogram** showing nested groupings.           | Excellent for **exploratory analysis** and visualization; allows you to choose $k$ by "cutting" the tree. Uses linkage methods (e.g., Average or Complete) with Bray-Curtis. |
| **K-means**                           | Partitions samples into $k$ groups defined by a calculated center (**centroid**). | **Avoid** unless data has been transformed (e.g., CLR) and Euclidean distance is appropriate, as it's sensitive to outliers and requires Euclidean geometry. |
| **Random Forest**                     | A **supervised classification** method.                      | **Not a clustering method** (it doesn't discover unknown groups). Used for predicting a known category (e.g., habitat type) based on ASV abundances. |

### Determine Optimal $k$

For PAM and HCA, the optimal number of clusters ($k$) can be determined using:

- **Silhouette Method:** Maximize the average silhouette score (closer to 1 is better).
- **Gap Statistic:** Compares within-cluster variation to random data.

------

# Python Implementation for Sample Clustering (QIIME 2 Artifacts)

## 1. Setup and Load Data

We use `qiime2` to load the feature table and `pandas` for manipulation.

Python

```
import qiime2 as q2
import pandas as pd
from scipy.spatial.distance import pdist, squareform
from sklearn_extra.cluster import KMedoids # For PAM
from sklearn.cluster import AgglomerativeClustering # For HCA
from scipy.cluster.hierarchy import linkage, dendrogram
from sklearn.metrics import silhouette_score
import matplotlib.pyplot as plt

# Load the QIIME 2 Feature Table
table_qza = q2.Artifact.load('path/to/your/table.qza')
feature_table_df = table_qza.view(pd.DataFrame)
```

## 2. Calculate Bray-Curtis Dissimilarity

This is the essential step for ecological data.

Python

```
# Calculate pairwise Bray-Curtis distances (condensed vector)
bray_curtis_vector = pdist(feature_table_df, metric='braycurtis')

# Convert the vector into a square, symmetric distance matrix
bray_curtis_matrix = squareform(bray_curtis_vector)
bray_curtis_df = pd.DataFrame(
    bray_curtis_matrix, 
    index=feature_table_df.index, 
    columns=feature_table_df.index
)
```

## 3. Partitioning Around Medoids (PAM)

We use the pre-calculated Bray-Curtis matrix with `KMedoids`.

Python

```
# --- A. Determine Optimal k using Silhouette Score ---
k_range = range(2, 6)
silhouette_scores = {}

for k in k_range:
    # Use 'precomputed' metric since we are passing the distance matrix
    kmedoids = KMedoids(n_clusters=k, metric='precomputed', random_state=42)
    kmedoids.fit(bray_curtis_matrix)
    
    score = silhouette_score(
        bray_curtis_matrix, 
        kmedoids.labels_, 
        metric='precomputed'
    )
    silhouette_scores[k] = score

optimal_k = max(silhouette_scores, key=silhouette_scores.get)
print(f"Optimal k (PAM): {optimal_k}")

# --- B. Apply Final PAM Clustering ---
final_kmedoids = KMedoids(n_clusters=optimal_k, metric='precomputed', random_state=42)
final_kmedoids.fit(bray_curtis_matrix)

pam_clusters = pd.Series(final_kmedoids.labels_, index=feature_table_df.index, name='PAM_Cluster')
print("\nPAM Cluster Assignments:")
print(pam_clusters.head())
```

## 4. Hierarchical Clustering (HCA)

HCA is useful for visualizing the relationships. **Average** linkage is appropriate for Bray-Curtis.

Python

```
# --- A. Perform Linkage (uses the condensed Bray-Curtis distance vector)
linked = linkage(bray_curtis_vector, method='average')

# --- B. Visualize the Dendrogram
plt.figure(figsize=(10, 7))
dendrogram(
    linked,
    orientation='top',
    labels=feature_table_df.index.tolist(),
    distance_sort='descending',
    show_leaf_counts=True
)
plt.title('Hierarchical Clustering (Bray-Curtis + Average Linkage)')
plt.ylabel('Bray-Curtis Dissimilarity Distance')
plt.xticks(rotation=90)
plt.tight_layout()
plt.show()
# 

# --- C. Cut the Dendrogram (e.g., using the optimal k from PAM)
hca = AgglomerativeClustering(
    n_clusters=optimal_k, 
    metric='precomputed', 
    linkage='average'
)
hca_clusters = pd.Series(hca.fit_predict(bray_curtis_matrix), index=feature_table_df.index, name='HCA_Cluster')
print("\nHCA Cluster Assignments:")
print(hca_clusters.head())
```
